notebooks/vision/horses_or_humans.livemd

```diff
@@ -0,0 +1,253 @@
+
+    {image, label}
+  end
+
+  defnp augment(images) do
+    # Convert images from NHWC to NCHW
+    images = Nx.transpose(images, axes: [0, 3, 1, 2])
+
+    # Normalize
+    images = images / 255.0
+
+    # Optional vertical/horizontal flip
+    u = Nx.random_uniform({})
+
+    cond do
+      u < 0.25 -> images
+      u < 0.5 -> Nx.reverse(images, axes: [2])
+      u < 0.75 -> Nx.reverse(images, axes: [3])
+      true -> Nx.reverse(images, axes: [2, 3])
+    end
+  end
+end
+```
+
+## Building the model
+
+The next step is creating our model. In this notebook, we choose the classic Convolutional Neural Network architecture. Let's dive in to the core components of a CNN.
+
+<!-- livebook:{"break_markdown":true} -->
+
+`Axon.conv/3` adds a convolutional layer, which is at the core of a CNN. A convolutional layer applies a filter function throughout the image, sliding a window with shape `:kernel_size`. As opposed to dense layers, a convolutional layer captures spacial information, because it works across multiple axes. This feature is a natural fit for images.
+
+<!-- livebook:{"break_markdown":true} -->
+
+| ![](https://miroslawmamczur.pl/wp-content/uploads/2021/03/06.gif)                       |
+| :-------------------------------------------------------------------------------------: |
+| Figure 1: A step-by-step visualization of a convolution layer for `kernel_size: {3, 3}` |
+
+<!-- livebook:{"break_markdown":true} -->
+
+`Axon.max_pool/2` adds a downscaling operation that takes the maximum value from a subtensor according to `:kernel_size`.
+
+<!-- livebook:{"break_markdown":true} -->
+
+| ![](https://production-media.paperswithcode.com/methods/MaxpoolSample2.png) |
+| :-------------------------------------------------------------------------: |
+| Figure 2: Max pooling operation for `kernel_size: {2, 2}`                   |
+
+<!-- livebook:{"break_markdown":true} -->
+
+`Axon.dropout/2` and `Axon.spatial_dropout/2` add dropout layers which prevent a neural network from overfitting. Standard dropout drops a given rate of randomly chosen neurons during the training process. On the other hand, spatial dropout gets rid of whole feature maps. The graphical difference between dropout and spatial dropout is presented in a picture below.
+
+<!-- livebook:{"break_markdown":true} -->
+
+| ![](https://miro.medium.com/max/1400/1*KkqxjvXTIV_b365B41ltfg.png)    |
+| :-------------------------------------------------------------------: |
+| Figure 3: The difference between standard dropout and spatial dropout |
+
+<!-- livebook:{"break_markdown":true} -->
+
+Knowing the relevant building blocks, let's build our network! It will have a convolutional part, composed of convolutional and pooling layers, this part should capture the spacial features of an image. Then at the end, we will add a dense layer with 512 neurons fed with all the spacial features, and a final two-neuron layer for as our classification output.
```

Comment:
```suggestion
Knowing the relevant building blocks, let's build our network! It will have a convolutional part, composed of convolutional and pooling layers, this part should capture the spatial features of an image. Then at the end, we will add a dense layer with 512 neurons fed with all the spatial features, and a final two-neuron layer for as our classification output.
```